README.md
webglobe: Interactive 3D Maps
=============================

You want to understand your data, but it's spatially distributed and you're
afraid that trying to make sense of it on something gross, like a Mercator
projection, is going to lead you to bad intuitions.

![Mercator Projection](vignettes/mercator.png)

(Greenland is nowhere near that big in reality.)

webglobe can help you do this! It allows you to interactively visualize your
data on either a three-dimensional globe or a flat map.



Example: Earth quakes
-----------------------------

    library(webglobe)              #Load the library
    data(quakes)                   #Load up some data

    wg <- webglobe(immediate=TRUE) #Make a webglobe (should open a net browser)
    Sys.sleep(10)                     #Wait for browser to start, or it won't work
    wg + wgpoints(quakes$lat, quakes$lon, size=5*quakes$mag) #Plot quakes
    wg + wgcamcenter(-24, 178.0650, 8000)                    #Move camera

![Webglobe earthquakes visualization](vignettes/webglobe_quakes.png)



Example: States
-----------------------------

    library(webglobe)                 #Load the library
    m  <- ggplot2::map_data("state")  #Get data
    m$extrude_height <- 1000000*runif(nrow(m),min=0,max=1)
    wg <- webglobe(immediate=TRUE)    #Make a webglobe (should open a net browser)
    Sys.sleep(10)                     #Wait for browser to start, or it won't work
    wg + wgpolygondf(m,fill="blue",alpha=1,stroke=NA)
    Sys.sleep(10)                     #Wait, in case you copy-pasted this
    wg + wgclear()                    #Clears the above

![Webglobe states visualization](vignettes/webglobe_states.png)



Modes
-----------------------------

webglobes have two modes: **immediate** and **not-immediate**. Immediate mode
displays a webglobe upon initialization and immediately prints all commands to
that globe. Not-immediate mode stores commands and displays them all at once,
allowing you to stage visualization without intermediate display. The difference
is illustrated below.

Display timing in intermediate mode:

    library(webglobe)
    data(quakes)                     #Get data
    q   <- quakes                    #Alias data
    wgi <- webglobe(immediate=TRUE)  #Webglobe is displayed now
    Sys.sleep(10)                    #Ensure webglobe runs before continuing
    wgi + wgpoints( q$lat,  q$lon)    #Data displays now!
    wgi + wgpoints(-q$lat, -q$lon)    #Data displays now!
    #Reloading the browser window clears everything

Display timing in not-intermediate mode:

    library(webglobe)
    data(quakes)                                  #Get data
    q   <- quakes                                 #Alias data
    wgn <- webglobe(immediate=FALSE)              #Webglobe is not displayed
    Sys.sleep(0)                                  #No need to wait
    #Note that we have to store commands
    wgn <- wgn + wgpoints( q$lat,  q$lon)         #Nothing shown yet
    wgn <- wgn + wgpoints(-q$lat, -q$lon)         #Nothing shown yet
    wgn <- wgn + wgcamcenter(2.89,-175.962,21460) #Nothing shown yet
    wgn                                           #Show it all now!
    #Reloading the browser window keeps everything

You can also switch between modes:

    library(webglobe)
    data(quakes)                                  #Get data
    q   <- quakes                                 #Alias data
    wgn <- webglobe(immediate=FALSE)              #Webglobe is not displayed
    Sys.sleep(0)                                  #No need to wait
    #Note that we have to store commands
    wgn <- wgn + wgpoints( q$lat,  q$lon)         #Nothing shown yet
    wgn <- wgn + wgpoints(-q$lat, -q$lon)         #Nothing shown yet
    wgn <- wgn + wgcamcenter(2.89,-175.962,21460) #Nothing shown yet
    wgn + wgimmediate()                           #Make it all immediate
    wgn
    wgn + wgpoints(q$lat, -q$lon)                 #This is shown right away
    #Reloading the browser window keeps everything up to `wgimmediate()`



Installation
-----------------------------

webglobe **hopefully will be** available from CRAN via:

    install.packages('webglobe')

If you want your code to be as up-to-date as possible, you can install it using:

    library(devtools) #Use `install.packages('devtools')` if need be
    install_github('r-barnes/webglobe', vignette=TRUE)



Developer Notes
-----------------------------

**How To Add Functionality**

There are really only two files that are import to contributing developers:
[inst/client/webglobe.js](inst/client/webglobe.js)
and
[R/webglobe.R](R/webglobe.R)
.

The package uses a JSON websocket message passing scheme to communicate data
between R and the JavaScript client.

Each `wg*()` function generates a JSON payload as follows:

    toString(jsonlite::toJSON(list(
      command = jsonlite::unbox("COMMAND_NAME"), #Required
      lat     = lat,                             #Example
      lon     = lon                              #Example
    )))

The payload consists of a `command` and accompanying data.

For more complex data, `geojsonio` can be leveraged to conveniently encode the
data. However, the resulting GeoJSON must be decoded, so that the whole packae
can be sent with only one level of encoding, as follows:

    toString(jsonlite::toJSON(list(
      command        = jsonlite::unbox("polygons"),
      polys          = jsonlite::fromJSON(geojsonio::geojson_json(df, group='group', geometry='polygon'))
    )))

On the JavaScript side, look for an object named `router`. `router` contains a
variety of fields which correspond to command names. To add a new command, add a
field with a corresponding function, such as:

    points: function(msg){
      var points = viewer.scene.primitives.add(new Cesium.PointPrimitiveCollection());
      for(var i=0;i<msg.lat.length;i++){
        points.add({
          position:  new Cesium.Cartesian3.fromDegrees(msg.lon[i], msg.lat[i], msg.alt[i]),
          color:     Cesium.Color[msg.colour[i].toUpperCase()],
          pixelSize: msg.size[i]
        });
      }
    }

Note that it is standard for the package to accept arguments such as `color`,
`size`, `width`, and so on as having either one value or a number of values
equal to the number of input points, polygons, or lines. That is: you should be
able to set the property of the entire group at once or at an individual level.

Note that functions added to [R/webglobe.R](R/webglobe.R) should be accompanied
by help text an examples, see the existing functions for templates.
Documentation should then be regenerated using

    roxygen2::roxygenise()

Changes to the vignettes (e.g. [vignettes/webglobe.Rmd](vignettes/webglobe.Rmd))
can be built using:

    devtools::build_vignettes()

It is polite to ensure that everything's good by using:

    devtools::check()

Once you have added your function, documented it, added any pertinent
explanations to the vignettes, and checked it, submit a pull request!


Licensing
-----------------------------

This package uses the following libraries:

 * cesiumjs: Cesium is licensed under Apache v2  

This package, and all code and documentation not otherwise mentioned above
(essentially anything outside the `src/` directory of this package) are released
under the MIT (Expat) license, as stated in the `LICENSE` file. The `LICENCE`
file exists for use with CRAN.



Roadmap
-----------------------------

* Make not-intermediate mode work

* Additional graphics primitives

* Submission to CRAN



Credits
-----------------------------

This R package was developed by Richard Barnes (http://rbarnes.org).

It uses the Cesium WebGL virtual globe and map engine ([link](https://cesiumjs.org/)).


NEWS
webglobe v???? (Release date: ????)
================
*Set "NeedsCompilation" to No



webglobe v1.0.2 (Release date: 2017-06-01)
================

* Updated description per advice from CRAN folks



webglobe v1.0.1 (Release date: 2017-06-01)
================

* Updated description



webglobe v1.0.0 (Release date: 2017-05-31)
================

* Added `wgtitle()`
* Added `label` to `wgpoints()`
* Allow hex colours
* Print error messags to the console, where they are more helpful



webglobe v0.0.5 (Release date: 2017-04-23)
================

* Restore `approximateTerrainHeights.json` so we can see polygons



webglobe v0.0.4 (Release date: 2017-04-22)
================

* Fixed issue with generics printing (but really)
* Fixed Github-CRAN licensing issue using a `.gitignore` file



webglobe v0.0.3 (Release date: 2017-04-21)
================

* Fixed issue with generics printing



webglobe v0.0.2 (Release date: 2017-04-21)
================

* Added unimmediate mode
* Added wgbar(), wgimmediate(), wgimmediate_set(), wgunimmediate()
* Display camera coordinates in viewer



webglobe v0.0.1 (Release date: 2017-04-20)
================

* Test release


inst/client/webglobe.js
//"use strict";
/*jshint browser: true */
/*global Cesium: false */
/*global console: false */

function getSceneCenter(){
  var lat = null;
  var lon = null;
  var alt = null;
  if (viewer.scene.mode == 3) {
    var windowPosition           = new Cesium.Cartesian2(viewer.container.clientWidth / 2, viewer.container.clientHeight / 2);
    var pickRay                  = viewer.scene.camera.getPickRay(windowPosition);
    var pickPosition             = viewer.scene.globe.pick(pickRay, viewer.scene);
    var pickPositionCartographic = viewer.scene.globe.ellipsoid.cartesianToCartographic(pickPosition);
    lon                          = pickPositionCartographic.longitude * (180 / Math.PI);
    lat                          = pickPositionCartographic.latitude * (180 / Math.PI);
    alt                          = viewer.camera.getMagnitude();
  } else if (viewer.scene.mode == 2) {
    var camPos = viewer.camera.positionCartographic;
    lat        = camPos.latitude * (180 / Math.PI);
    lon        = camPos.longitude * (180 / Math.PI);
    alt        = viewer.camera.getMagnitude();
  }
  return [lat, lon, alt];
}

var imageryViewModels = [];

imageryViewModels.push(new Cesium.ProviderViewModel({
  name : 'Open Street Map Offline',
  iconUrl : Cesium.buildModuleUrl('Widgets/Images/ImageryProviders/openStreetMap.png'),
  tooltip : 'OpenStreetMap (OSM) is a collaborative project to create a free editable map of the world.\nhttp://www.openstreetmap.org',
  creationFunction : function() {
    return new Cesium.createOpenStreetMapImageryProvider({
      url : 'tiles/osm',
      minimumLevel: 0,
      maximumLevel: 4
    });
  }
}));

imageryViewModels.push(new Cesium.ProviderViewModel({
  name : 'Open\u00adStreet\u00adMap',
  iconUrl : Cesium.buildModuleUrl('Widgets/Images/ImageryProviders/openStreetMap.png'),
  tooltip : 'OpenStreetMap (OSM) is a collaborative project to create a free editable map of the world.\nhttp://www.openstreetmap.org',
  creationFunction : function() {
    return new Cesium.createOpenStreetMapImageryProvider({
      url : '//a.tile.openstreetmap.org/'
    });
  }
}));

var viewer_options = {
    animation:                              false,
    fullscreenButton:                       true,
    geocoder:                               false,
    homeButton:                             true,
    infoBox:                                false,
    sceneModePicker:                        true,
    selectionIndicator:                     false,
    timeline:                               false,
    navigationHelpButton:                   false,
    navigationInstructionsInitiallyVisible: false,
    scene3DOnly:                            false,
    skyBox:                                 false,
    skyAtmosphere:                          false,
    sceneMode:                              Cesium.SceneMode.SCENE3D,
    baseLayerPicker:                        true,
    imageryProviderViewModels:              imageryViewModels,
    terrainProviderViewModels:              [],
    // imageryProvider:                     new Cesium.createOpenStreetMapImageryProvider({
    //   url:                               'https://a.tile.openstreetmap.org/'
    // }),

    // terrainProvider : new Cesium.CesiumTerrainProvider({
    //   url : '//assets.agi.com/stk-terrain/world'
    // }),
    targetFrameRate: 100,
    orderIndependentTranslucency: false,
    contextOptions: {
      webgl : {
        alpha:                        false,
        depth:                        false,
        stencil:                      false,
        antialias:                    false,
        premultipliedAlpha:           true,
        preserveDrawingBuffer:        false,
        failIfMajorPerformanceCaveat: false
      },
      allowTextureFilterAnisotropic : false
    }
};










var router = {
  clear: function(msg){
    viewer.dataSources.removeAll();
    viewer.scene.primitives.removeAll();
  },
  polygons: function(msg){
    var promise = Cesium.GeoJsonDataSource.load(msg.polys);
    promise.then(function(dataSource) {
      viewer.dataSources.add(dataSource);

      var entities = dataSource.entities.values;
      for (var i = 0; i < entities.length; i++) {
        var entity = entities[i];

        var alpha = msg.alpha;
        if(entity.properties.alpha)
          alpha = entity.properties.alpha;


        if(entity.properties.fill){
          entity.polygon.material = Cesium.Color.fromCssColorString(entity.properties.fill.valueOf()).withAlpha(alpha);
        } else if(msg.fill){
          entity.polygon.material = Cesium.Color.fromCssColorString(msg.fill).withAlpha(alpha);
        } else {
          entity.polygon.material = Cesium.Color.TRANSPARENT;
        }

        console.log(entity);

        //TODO
        if(entity.properties.stroke){
          entity.polygon.stroke = Cesium.Color.fromCssColorString(entity.properties.stroke.valueOf()).withAlpha(alpha);
        } else if(msg.stroke){
          entity.polygon.stroke = Cesium.Color.fromCssColorString(msg.stroke).withAlpha(alpha);
        } else {
          entity.polygon.outline = false;
        }

        //TODO
        if(entity.properties.stroke_width){
          entity.polygon.strokeWidth = parseFloat(entity.properties.stroke_width);
        } else if(msg.stroke_width){
          entity.polygon.strokeWidth = msg.stroke_width;
        } else {
          entity.polygon.outline = false;
        }

        if(entity.properties.extrude_height)
          entity.polygon.extrudedHeight = parseFloat(entity.properties.extrude_height);
        else if(msg.extrude_height)
          entity.polygon.extrudedHeight = msg.extrude_height;
      }
    }).otherwise(function(error){
      //Display any errrors encountered while loading.
      console.error(error);
    });
  },
  points: function(msg){
    for(var i=0;i<msg.lat.length;i++){
      var newpoint = {
        position: Cesium.Cartesian3.fromDegrees(msg.lon[i], msg.lat[i], msg.alt[i]),
        point:    {
          pixelSize: msg.size[i],
          color:     Cesium.Color.fromCssColorString(msg.colour[i])
        }
      };

      if(msg.label[i]){
        newpoint.label = {
          text:           msg.label[i],
          font:           '14pt monospace',
          style:          Cesium.LabelStyle.FILL_AND_OUTLINE,
          outlineWidth:   2,
          verticalOrigin: Cesium.VerticalOrigin.BOTTOM,
          pixelOffset:    new Cesium.Cartesian2(0, -9)
        };
      }

      viewer.entities.add(newpoint);
    }
  },
  bars: function(msg){
    for(var i=0;i<msg.lat.length;i++){
      viewer.entities.add({
        name:     'bar',
        polyline: {
          positions : Cesium.Cartesian3.fromDegreesArrayHeights(
            [msg.lon[i], msg.lat[i], 0, msg.lon[i], msg.lat[i], msg.alt[i]]
          ),
          width:    msg.width[i],
          material: new Cesium.PolylineOutlineMaterialProperty({
            color : Cesium.Color.fromCssColorString(msg.colour[i])
            //outlineWidth : 2,
            //outlineColor : Cesium.Color.BLACK
          })
        }
      });
    }
  },
  cam_reset:  function(msg){
    viewer.camera.flyHome();
  },
  cam_center: function(msg){
    if(msg.alt===null)
      msg.alt = viewer.camera.positionCartographic();
    viewer.camera.flyTo({destination:Cesium.Cartesian3.fromDegrees(msg.lon,msg.lat,msg.alt)});
  },
  title: function(msg){
    document.title = msg.title;
  }
};

var viewer   = new Cesium.Viewer('webglobe', viewer_options);
var ws       = new WebSocket("ws://"+window.location.host);

ws.onmessage = function(msg){
  msg = JSON.parse(msg.data);
  console.log(msg);
  router[msg.command](msg);
};

ws.onopen = function(e){
  ws.send('sally_forth');
};

var pos_interval_handle = setInterval(function() {
  var pos = viewer.camera.positionCartographic;
  var lat = pos.latitude *180/Math.PI;
  var lon = pos.longitude*180/Math.PI;
  var alt = pos.height   /1000;
  
  document.getElementById('currentpos').innerHTML = 
    lat.toFixed(5) + "&deg;, " + lon.toFixed(5) + "&deg;, " + alt.toFixed(0);
}, 1000);


R/webglobe.R
#' @importFrom geojsonio  geojson_json
#' @importFrom httpuv     stopDaemonizedServer startDaemonizedServer
#' @importFrom jsonlite   toJSON fromJSON unbox
#' @importFrom stats      runif
#' @importFrom utils      browseURL

content_types <- list(
  js   = "application/javascript",
  html = "text/html",
  css  = "text/css",
  png  = "image/png", 
  jpg  = "image/jpeg",
  json = "application/json"
)

wg404 <- list(
  status  = 404L,
  headers = list("Content-Type" = "text/plain"),
  body    = "Not found"
)

wg403 <- list(
  status  = 403L,
  headers = list("Content-Type" = "text/plain"),
  body    = "Forbidden"
)

wg400 <- list(
  status  = 400L,
  headers = list("Content-Type" = "text/plain"),
  body    = "Bad request"
)

wgstopbyenv <- function(wgenv){
  httpuv::stopDaemonizedServer(wgenv[['server']])
}

wgsend <- function(wg,msg){
  if(wg$env[['immediate']]){
    wgsendsend(wg,msg)
  } else {
    wg$env[['msgs']] <- c(wg$env[['msgs']],msg)
  }
}

wgsendsend <- function(wg,msg){
  tryCatch({
    wg$env[['ws']]$send(msg)
  }, error=function(e){
    print('Error: Could not send to client. Make sure you have a browser open!')
  })
}



#' @name print.webglobe
#' 
#' @title      Display a webglobe
#'
#' @description
#'             Displays a webglobe. If the webglobe is immediate, then a browser
#'             window containing it should already be open; in this case, the
#'             webglobe's address is returned. If the webglobe is not immediate
#'             then a new browser is open and the cached pipeline is sent to it.
#' 
#' @param x    The webglobe   
#' @param ...  Ignored 
#'
#' @return     NA
#'
#' @examples 
#' \dontrun{
#' library(webglobe)
#' wg<-webglobe()
#' wg
#' }
#'
#' @export 
print.webglobe <- function(x, ...){
  if(x$env[['immediate']])
    print(paste0("Server should be running at 'http://localhost:",x$env[['port']],"'"))
  else
    utils::browseURL(paste0('http://localhost:',x$env[['port']]))
  NA
}

#' @name is.webglobe
#' 
#' @title      Is it a webglobe?
#'
#' @description
#'             Determine if an object is a webglobe
#' 
#' @param x    The object that might be a webglobe
#'
#' @return     TRUE or FALSE
#'
#' @examples 
#' \dontrun{
#' library(webglobe)
#' wg<-webglobe(immediate=TRUE)
#' is.webglobe(wg)
#' }
#'
#' @export 
is.webglobe <- function(x) inherits(x, "webglobe")

#' @name +.webglobe
#' 
#' @title      Send command
#'
#' @description
#'             Send a command to a webglobe
#' 
#' @param wg   Webglobe
#' @param x    Command to send
#'
#' @return     The same webglobe
#'
#' @examples 
#' \dontrun{
#' library(webglobe)
#' wg<-webglobe(immediate=TRUE)
#' wg + wgclear()
#' }
#'
#' @export 
`+.webglobe` <- function(wg,x){
  if(x=="immediate")
    wg$env[['immediate']] <- TRUE
  else if (x=="unimmediate")
    wg$env[['immediate']] <- FALSE
  else
    wgsend(wg,x)
  wg
}


#' @name webglobe
#' 
#' @title      Make a new webglobe
#'
#' @description
#'             Constructs a new webglobe and starts its server
#' 
#' @param immediate 
#'             Whether the webglobe should immediately show the results of
#'             graphics commands or additively cache them. `immediate` mode can
#'             be used to experimentally build up a pipeline. Once established
#'             this can be stored in a non-immediate webglobe for easy acces
#'             later
#'
#' @return     A webglobe object
#'
#' @examples 
#' \dontrun{
#' library(webglobe)
#' wg<-webglobe(immediate=TRUE)
#' }
#'
#' @export 
webglobe <- function(immediate=FALSE){
  the_env <- new.env(parent=emptyenv())
  app <- list(
    call = function(req) {
      if (!identical(req$REQUEST_METHOD, 'GET'))
        return(NULL)

      path <- req$PATH_INFO

      if (is.null(path))
        return(wg400)

      if (path == '/')
        path <- '/index.html'

      path <- file.path(path.package('webglobe'),'client',path)

      ctype <- content_types[[tools::file_ext(path)]]
      if(is.null(ctype))
        return(wg403)

      if(!file.exists(path))
        return(wg404)

      fcontents <- readBin(path, 'raw', n=file.info(path)$size)
      return(list(status=200L, headers=list('Content-Type'=ctype), body=fcontents))
    },
    onWSOpen = function(ws) {
      the_env[['ws']]<-ws
      ws$onMessage(function(binary, message) {
        if(message!="sally_forth")
          return
        for(m in the_env[['msgs']])
          ws$send(m)
      })
    },
    env = the_env
  )

  app$env[['msgs']]      <- c()
  app$env[['immediate']] <- immediate

  startServer <- function(depth){
    tryCatch({
      app$env[['port']]   <- floor(stats::runif(1,min=4000,max=8000))
      app$env[['server']] <- httpuv::startDaemonizedServer("0.0.0.0", app$env[['port']], app)
      TRUE
    }, error = function(e) {
      if(depth==100){
        return(FALSE)
      } else {
        return(startServer(depth+1))
      }
    })
  }

  if(!startServer(0)){
    stop('Could not start the server - probably no port was available.')
  }

  class(app) <- "webglobe"

  reg.finalizer(app$env, wgstopbyenv, onexit = TRUE)

  if(immediate)
  utils::browseURL(paste0('http://localhost:',app$env[['port']]))

  return(app)
}



#' @name wgport
#' 
#' @title      Get webglobe's port
#'
#' @description
#'             Determine which port a webglobe is running on
#'
#' @param wg   Webglobe whose port should be determined
#' 
#' @return     A number representing the webglobe's port
#'
#' @examples 
#' \dontrun{
#' library(webglobe)
#' wg<-webglobe(immediate=TRUE)
#' wgport(webglobe)
#' }
#'
#' @export 
wgport <- function(wg){
  wg$env[['port']]
}

#' @name wgpoints
#' 
#' @title      Plot points
#'
#' @description
#'             Plots latitude-longitude points
#'
#' @param lat    One or more latitude values
#' @param lon    One or more longitude values
#' @param label  Label to put next to point
#' @param alt    Altitude of the points, can be single value or vector
#' @param colour Colour name of the points, can be single value or vector
#' @param size   Size of the points, can be single value or vector
#' 
#' @return     A webglobe command
#'
#' @examples 
#' \dontrun{
#' library(webglobe)
#' wg <- webglobe(immediate=FALSE)
#' wg <- wg + wgpoints(c(45,20),c(-93,127),alt=3,colour=c("blue","red"))
#' wg <- wg + wgpoints(51.5074,-0.1278,label="London",alt=0,colour="blue")
#' wg
#' }
#'
#' @export 
wgpoints <- function(lat,lon,label=NA,alt=0,colour="yellow",size=10){
  if(length(lat)!=length(lon))
    stop('Same number of latitude and longitude points are required!')
  if(length(alt)!=1 && length(alt)!=length(lat))
    stop('One altitude must be specified, or a number equal to that of latitude/longitude!')
  if(length(alt)==1)
    alt <- rep(alt, length(lat))
  if(length(colour)==1)
    colour <- rep(colour, length(lat))
  if(length(size)==1)
    size <- rep(size, length(lat))
  if(length(label)==1)
    label <- rep(label, length(lat))
  toString(jsonlite::toJSON(list(
    command = jsonlite::unbox("points"),
    lat     = lat,
    lon     = lon,
    alt     = alt,
    colour  = colour,
    size    = size,
    label   = label
  )))
}

#' @name wgcamhome
#' 
#' @title      Camera: Send home
#'
#' @description
#'             Send the camera to its home location
#' 
#' @return     A webglobe command
#'
#' @examples 
#' \dontrun{
#' library(webglobe)
#' wg<-webglobe(immediate=TRUE)
#' wg+wgcamhome()
#' }
#'
#' @export 
wgcamhome <- function(){
  toString(jsonlite::toJSON(list(
    command = jsonlite::unbox("cam_reset")
  )))
}

#' @name wgclear
#' 
#' @title      Clear the scene
#'
#' @description
#'             Clears everything from the map
#' 
#' @return     A webglobe command
#'
#' @examples 
#' \dontrun{
#' library(webglobe)
#' wg<-webglobe(immediate=TRUE)
#' wg+wgclear()
#' }
#'
#' @export 
wgclear <- function(){
  toString(jsonlite::toJSON(list(
    command = jsonlite::unbox("clear")
  )))
}

#' @name wgcamcenter
#' 
#' @title      Camera: Center on a point
#'
#' @description
#'             Centers the camera on a point
#'
#' @param lat  Latitude of the center point, in degrees
#' @param lon  Longitude of the center point, in degrees
#' @param alt  Altitude of the center point, in kilometres
#' 
#' @return     A webglobe command
#'
#' @examples 
#' \dontrun{
#' library(webglobe)
#' wg<-webglobe(immediate=TRUE)
#' wg+wgcamcenter(45,-93,5000)
#' }
#'
#' @export 
wgcamcenter <- function(lat,lon,alt=NA){
  if(!is.na(alt))
    alt<-1000*alt
  toString(jsonlite::toJSON(list(
    command = jsonlite::unbox("cam_center"),
    lat     = jsonlite::unbox(lat),
    lon     = jsonlite::unbox(lon),
    alt     = jsonlite::unbox(alt)
  )))
}

#' @name wgpolygondf
#' 
#' @title      Plot long-frame polygons
#'
#' @description
#'             Plot polygons defined by long-style data frame
#'
#' @param df              Data frame to plot
#' @param fill            Fill colour name
#' @param alpha           Alpha (transparency value)
#' @param extrude_height  Height of the polygon above the surrounding landscape, in TODO
#' @param stroke          Outline colour (TODO)
#' @param stroke_width    Outline width (TODO)
#' 
#' @return     A webglobe command
#'
#' @examples 
#' \dontrun{
#' library(webglobe)
#' wg<-webglobe(immediate=TRUE)
#' wg+wgpolygondf(ggplot2::map_data("usa"),fill="blue",extrude_height=1000)
#' }
#'
#' @export 
wgpolygondf <- function(df,fill=NA,alpha=1,extrude_height=0,stroke="yellow",stroke_width=10){
  toString(jsonlite::toJSON(list(
    command        = jsonlite::unbox("polygons"),
    polys          = jsonlite::fromJSON(geojsonio::geojson_json(df, group='group', geometry='polygon')),
    fill           = jsonlite::unbox(fill),
    extrude_height = jsonlite::unbox(extrude_height),
    alpha          = jsonlite::unbox(alpha),
    stroke         = jsonlite::unbox(stroke),
    stroke_width   = jsonlite::unbox(stroke_width)
  )))
}



#' @name wgtitle
#' 
#' @title      Title of webglobe browser window
#'
#' @description
#'             Changes the tab/window title of the webglobe's browser view
#'
#' @param title  The title to use
#' 
#' @return     A webglobe command
#'
#' @examples 
#' \dontrun{
#' library(webglobe)
#' wg<-webglobe(immediate=TRUE)
#' wg+wgtitle("I am the new title!")
#' }
#'
#' @export 
wgtitle <- function(title){
  toString(jsonlite::toJSON(list(
    command = jsonlite::unbox("title"),
    title   = jsonlite::unbox(title)
  )))
}



#' @name wgbar
#' 
#' @title      Plot bars from the surface
#'
#' @description
#'             Plots bars rising upwards from points on the Earth's surface
#'
#' @param lat    Latitude of the bars' bases, in degrees
#' @param lon    Latitude of the bars' bases, in degrees
#' @param alt    Altitude of the bars' tops, may be one or many values
#' @param colour Colour of the bars, may be one or many values
#' @param width  Width of bar bars, may be one or many values
#' 
#' @return     A webglobe command
#'
#' @examples 
#' \dontrun{
#' library(webglobe)
#' data(quakes)                                                      #Load up some data
#' wg <- webglobe(immediate=FALSE)                                   #Make a webglobe
#' wg <- wg + wgbar(quakes$lat, quakes$lon, alt=1.5e6*quakes$mag/10) #Plot quakes
#' wg <- wg + wgcamcenter(-33.35, 142.96, 8000)                      #Move camera
#' wg
#' }
#'
#' @export 
wgbar <- function(lat,lon,alt=3000000,colour="blue",width=3){
  if(length(lat)!=length(lon))
    stop('Same number of latitude and longitude points are required!')
  if(length(alt)!=1 && length(alt)!=length(lat))
    stop('One altitude must be specified, or a number equal to that of latitude/longitude!')
  if(length(colour)!=1 && length(colour)!=length(lat))
    stop('One colour must be specified, or a number equal to that of latitude/longitude!')
  if(length(width)!=1 && length(width)!=length(lat))
    stop('One width must be specified, or a number equal to that of latitude/longitude!')
  if(length(alt)==1)
    alt <- rep(alt, length(lat))
  if(length(colour)==1)
    colour <- rep(colour, length(lat))
  if(length(width)==1)
    width <- rep(width, length(lat))
  toString(jsonlite::toJSON(list(
    command = jsonlite::unbox("bars"),
    lat     = lat,
    lon     = lon,
    alt     = alt,
    colour  = colour,
    width   = width
  )))
}



#' @name wgimmediate
#' 
#' @title      Immediate mode: On
#'
#' @description
#'             Turns on immediate mode
#' 
#' @return     A webglobe command
#'
#' @examples 
#' \dontrun{
#' library(webglobe)
#' wg<-webglobe(immediate=FALSE)
#' wg + wgimmediate() #wg is now immediate
#' }
#'
#' @export 
wgimmediate <- function(){
  "immediate"
}

#' @name wgunimmediate
#' 
#' @title      Immediate mode: Off
#'
#' @description
#'             Turns off immediate mode
#' 
#' @return     A webglobe command
#'
#' @examples 
#' \dontrun{
#' library(webglobe)
#' wg<-webglobe(immediate=TRUE)
#' wg + wgunimmediate() #wg is now unimmediate
#' }
#'
#' @export 
wgunimmediate <- function(){
  "unimmediate"
}

#' @name wgimmediate_set
#' 
#' @title      Immediate mode: Set
#'
#' @description
#'             Set immediate mode by value
#'
#' @param mode TRUE or FALSE: TRUE immplies immediate mode on, FALSE implies off
#' 
#' @return     A webglobe command
#'
#' @examples 
#' \dontrun{
#' library(webglobe)
#' wg<-webglobe(immediate=TRUE)
#' wg + wgimmediate_set(FALSE) #wg is now unimmediate
#' }
#'
#' @export 
wgimmediate_set <- function(mode){
  if(mode)
    "immediate"
  else
    "unimmediate"
}
